src/hotspot/share/utilities/debug.cpp

```diff
@@ -349,8 +349,26 @@ void report_java_out_of_memory(const char* message) {
   // commands multiple times we just do it once when the first threads reports
   // the error.
   if (Atomic::cmpxchg(&out_of_memory_reported, 0, 1) == 0) {
+
+    // SapMachine 2021-05-21: On SapMachine, in contrast to upstream OpenJDK, this
+    //  function gets invoked for OOMs resulting from Thread resource exhaustion as
+    //  well as for the "normal" Heap/Metaspace exhaustion. However, in that case we
+    //  do not handle HeapDumpOnOutOfMemoryError since heap dumps may take a long time
+    //  and are probably not as useful.
+    // (Note: the strstr is a bit tacky here but keeps the patch small.)
+    const bool is_thread_exhaustion =
+        ::strstr(message, "unable to create native thread") != NULL;
+
+    // SapMachine 2021-05-21: Print stack to stdout
+    if ((HeapDumpOnOutOfMemoryError && !is_thread_exhaustion) ||
+        (OnOutOfMemoryError && OnOutOfMemoryError[0]) ||
+        CrashOnOutOfMemoryError || ExitOnOutOfMemoryError) {
+      // SapMachine 2021-05-21: Print stack to stdout
+      VMError::print_stack(tty);
+    }
+
     // create heap dump before OnOutOfMemoryError commands are executed
-    if (HeapDumpOnOutOfMemoryError) {
+    if (HeapDumpOnOutOfMemoryError && !is_thread_exhaustion) {
```

Comment:
I think it would be best to always get the thread dump when we have thread exhaustion and it should be printed before we try a heap dump. Dumping the heap after dumping the threads is OK, since the most important information is already dumped.